Index: util/grub-probe.c
===================================================================
--- util/grub-probe.c	(revision 2571)
+++ util/grub-probe.c	(working copy)
@@ -239,6 +239,9 @@ probe (const char *path, char *device_na
 
       stat (path, &st);
 
+      /* FIXME: `path' can't be used to read a file via GRUB facilities,
+         because it's not relative to its root.  */
+#if 0
       if (S_ISREG (st.st_mode))
 	{
 	  /* Regular file.  Verify that we can read it properly.  */
@@ -258,6 +261,8 @@ probe (const char *path, char *device_na
 	  if (memcmp (filebuf_via_grub, filebuf_via_sys, file->size))
 	    grub_util_error ("files differ");
 	}
+#endif
+
       printf ("%s\n", fs->name);
     }
 
