diff -ur grub2-1.96+20080724/normal/main.c grub2-1.96+20080724-2/normal/main.c
--- grub2-1.96+20080724/normal/main.c	2008-02-02 17:48:52.000000000 +0100
+++ grub2-1.96+20080724-2/normal/main.c	2008-07-29 17:19:46.024867049 +0200
@@ -97,9 +97,6 @@
 	}
       else
 	{
-	  if (c == '\n')
-	    break;
-
 	  if (pos >= max_len)
 	    {
 	      char *old_cmdline = cmdline;
@@ -112,6 +109,9 @@
 		}
 	    }
 
+         if (c == '\n')
+           break;
+
 	  cmdline[pos++] = c;
 	}
     }
