Index: kern/i386/pc/init.c
===================================================================
--- kern/i386/pc/init.c	(Revision 1763)
+++ kern/i386/pc/init.c	(Arbeitskopie)
@@ -64,7 +64,7 @@
   /* XXX: This should be enough.  */
   char dev[100];
 
-  if (grub_install_dos_part != -2)
+  if (grub_prefix[0] != '(')
     {
       /* If the root drive is not set explicitly, assume that it is identical
          to the boot drive.  */
