--- grub2-1.96+20080724.orig/disk/raid.c	2008-07-04 00:56:43.000000000 +0200
+++ grub2-1.96+20080724/disk/raid.c	2008-08-12 14:09:32.000000000 +0200
@@ -175,11 +175,11 @@
 	    err = grub_disk_read (array->device[disknr], read_sector, 0,
 				  read_size << GRUB_DISK_SECTOR_BITS, buf);
 	    if (err)
 	      break;
 
-	    buf += read_size;
+	    buf += (read_size << GRUB_DISK_SECTOR_BITS);
 	    size -= read_size;
 	    if (! size)
 	      break;
 
 	    if (size > array->chunk_size)
