diff -Nur grub2-1.94/conf/powerpc-ieee1275.rmk grub2-1.94.new/conf/powerpc-ieee1275.rmk
--- grub2-1.94/conf/powerpc-ieee1275.rmk	2006-05-28 20:01:43.000000000 -0300
+++ grub2-1.94.new/conf/powerpc-ieee1275.rmk	2006-06-07 21:32:59.000000000 -0300
@@ -26,9 +26,9 @@
 pkgdata_PROGRAMS = kernel.elf
 
 # Utilities.
-bin_UTILITIES = grub-emu
-sbin_UTILITIES = grub-mkimage
-
+bin_UTILITIES = grub-mkimage
+sbin_UTILITIES = grub-emu 
+ 
 # For grub-mkimage.
 grub_mkimage_SOURCES = util/powerpc/ieee1275/grub-mkimage.c util/misc.c \
         util/resolve.c 
