--- util/biosdisk.c	2008-02-11 16:47:35.000000000 +0100
+++ util/biosdisk.c	2008-02-28 17:00:20.000000000 +0100
@@ -584,7 +584,7 @@
   sprintf (p, (drive & 0x80) ? "hd%d" : "fd%d", drive & ~0x80);
   
   if (dos_part >= 0)
-    sprintf (p + strlen (p), ",%d", dos_part + 1);
+    sprintf (p + strlen (p), ",%d", dos_part + (getenv ("GRUB_LEGACY_0_BASED_PARTITIONS") ? 0 : 1));
   
   if (bsd_part >= 0)
     sprintf (p + strlen (p), ",%c", bsd_part + 'a');
