--- disk/raid.c	2008-08-25 20:21:55.000000000 +0200
+++ disk/raid.c	2008-08-25 20:23:28.000000000 +0200
@@ -459,7 +459,7 @@
       array->version = sb.major_version;
       array->level = sb.level;
       array->layout = sb.layout;
-      array->total_devs = sb.nr_disks;
+      array->total_devs = sb.raid_disks;
       array->nr_devs = 0;
       array->uuid[0] = sb.set_uuid0;
       array->uuid[1] = sb.set_uuid1;
