
Author: Robert Millan

An ugly kludge.  Should this be merged upstream?

--- util/biosdisk.c	2008-04-07 18:45:01.000000000 +0200
+++ util/biosdisk.c	2008-04-10 10:52:32.000000000 +0200
@@ -550,6 +550,16 @@ read_device_map (const char *dev_map)
 	  continue;
 	}
 
+      if (! strncmp (p, "/dev/fd", sizeof ("/dev/fd") - 1))
+	{
+	  char *q = p + sizeof ("/dev/fd") - 1;
+	  if (*q >= '0' && *q <= '9')
+	    {
+	      grub_util_info ("`%s' looks like a floppy drive, skipping", p);
+	      continue;
+	    }
+	}
+
 #ifdef __linux__
       /* On Linux, the devfs uses symbolic links horribly, and that
 	 confuses the interface very much, so use realpath to expand
