diff -Nur -x '*.orig' -x '*~' grub2-1.96+20080724/util/grub.d/10_linux.in grub2-1.96+20080724.new/util/grub.d/10_linux.in
--- grub2-1.96+20080724/util/grub.d/10_linux.in	2008-06-26 17:03:26.000000000 +1000
+++ grub2-1.96+20080724.new/util/grub.d/10_linux.in	2008-11-17 11:25:53.000000000 +1100
@@ -24,7 +24,7 @@
 if [ "x${GRUB_DISTRIBUTOR}" = "x" ] ; then
   OS=GNU/Linux
 else
-  OS="${GRUB_DISTRIBUTOR} GNU/Linux"
+  OS="${GRUB_DISTRIBUTOR}"
 fi
 
 # loop-AES arranges things so that /dev/loop/X can be our root device, but
