diff -Nur -x '*.orig' -x '*~' grub2-1.96+20080512/util/update-grub.in grub2-1.96+20080512.new/util/update-grub.in
--- grub2-1.96+20080512/util/update-grub.in	2008-04-30 06:06:26.000000000 -0500
+++ grub2-1.96+20080512.new/util/update-grub.in	2008-05-28 09:34:36.000000000 -0500
@@ -102,6 +102,8 @@
 
 # Device containing our userland.  Typically used for root= parameter.
 GRUB_DEVICE="`grub-probe --target=device /`"
+#Convert this to a UUID
+GRUB_DEVICE="`convert_root_to_uuid $GRUB_DEVICE`"
 
 # Filesystem for the device containing our userland.  Used for stuff like
 # choosing Hurd filesystem module.
diff -Nur -x '*.orig' -x '*~' grub2-1.96+20080512/util/update-grub_lib.in grub2-1.96+20080512.new/util/update-grub_lib.in
--- grub2-1.96+20080512/util/update-grub_lib.in	2008-05-06 08:44:11.000000000 -0500
+++ grub2-1.96+20080512.new/util/update-grub_lib.in	2008-05-28 09:34:36.000000000 -0500
@@ -125,3 +125,42 @@
   fi
   return 0
 }
+
+convert_root_to_uuid()
+{
+	local root; root=$1
+
+	convert=false
+	root=$(echo "$root" | sed 's/.*root=//;s/ .*//')
+	case "$root" in
+		UUID=*|LABEL=*)
+			;;
+		/dev/disk/*)
+			;;
+		/dev/mapper/*)
+			;;
+		/dev/evms/[hs]d[a-z][0-9]*)
+			convert=:
+			;;
+		/dev/evms/*)
+			;;
+		/dev/md[0-9]*)
+			;;
+		/dev/*)
+			convert=:
+			;;
+	esac
+	if $convert; then
+		if [ -L "$DEV" ] && readlink "$DEV" | grep -q "^/dev/mapper/"
+		then
+			:
+		elif [ -b "$root" ]; then
+			uuid=$(/sbin/vol_id -u "$root" || true)
+			if [ -n "$uuid" ]; then
+				root="UUID=$uuid"
+			fi
+		fi
+	fi
+
+	echo "$root"
+}
