--- a/conf/powerpc-ieee1275.rmk
+++ b/conf/powerpc-ieee1275.rmk
@@ -16,7 +16,8 @@ DEFSYMFILES += kernel_syms.lst
 kernel_elf_HEADERS = boot.h cache.h device.h disk.h dl.h elf.h elfload.h \
 	env.h err.h file.h fs.h kernel.h misc.h mm.h net.h parser.h rescue.h \
 	symbol.h term.h time.h types.h powerpc/libgcc.h loader.h partition.h \
-	pc_partition.h ieee1275/ieee1275.h machine/kernel.h handler.h
+	pc_partition.h ieee1275/ieee1275.h machine/kernel.h handler.h list.h \
+	command.h
 
 symlist.c: $(addprefix include/grub/,$(kernel_elf_HEADERS)) config.h gensymlist.sh
 	/bin/sh gensymlist.sh $(filter %.h,$^) > $@ || (rm -f $@; exit 1)
@@ -87,8 +88,6 @@ kernel_elf_SOURCES = kern/powerpc/ieee1275/startup.S kern/ieee1275/cmain.c \
 	kern/parser.c kern/partition.c kern/env.c kern/powerpc/dl.c 	\
 	kern/generic/millisleep.c kern/time.c                            \
 	symlist.c kern/powerpc/cache.S
-kernel_elf_HEADERS = grub/powerpc/ieee1275/ieee1275.h list.h handler.h \
-	command.h
 kernel_elf_CFLAGS = $(COMMON_CFLAGS)
 kernel_elf_ASFLAGS = $(COMMON_ASFLAGS)
 kernel_elf_LDFLAGS = $(COMMON_LDFLAGS) -static-libgcc -lgcc \
