--- grub2-1.96+20080724.orig/util/getroot.c	2008-05-16 23:39:52.000000000 +0200
+++ grub2-1.96+20080724/util/getroot.c	2008-08-12 14:56:14.000000000 +0200
@@ -461,10 +461,16 @@
       else if (os_dev[7] >= '0' && os_dev[7] <= '9')
 	{
 	  memcpy (grub_dev, os_dev + 5, 7);
 	  grub_dev[7] = '\0';
 	}
+      else if (os_dev[7] == '/' && os_dev[8] >= '0' && os_dev[8] <= '9')
+	{
+	  memcpy (grub_dev, os_dev + 5, 2);
+	  memcpy (grub_dev + 2, os_dev + 8, 5);
+	  grub_dev[7] = '\0';
+	}
       else
 	grub_util_error ("Unknown kind of RAID device `%s'", os_dev);
 
       break;
 
