--- grub2-1.96+20080724.orig/util/getroot.c	2008-08-12 15:02:14.000000000 +0200
+++ grub2-1.96+20080724/util/getroot.c	2008-08-12 15:05:22.000000000 +0200
@@ -410,9 +410,22 @@
   switch (grub_util_get_dev_abstraction (os_dev))
     {
     case GRUB_DEV_ABSTRACTION_LVM:
-      grub_dev = xmalloc (strlen (os_dev) - 12 + 1);
+      {
+        unsigned char i, j, k, l;
 
-      strcpy (grub_dev, os_dev + 12);
+        j = sizeof ("/dev/mapper/") -1;
+        l = strlen (os_dev) - j + 1;
+
+        grub_dev = xmalloc (strlen (os_dev) - strlen ("/dev/mapper/") + 1);
+
+        for (i = 0, k = 0; i < l; i++)
+          {
+             grub_dev[k] = os_dev[j + i];
+             k++;
+             if (os_dev[j + i] == '-' && os_dev[j + i + 1] == '-')
+               i++;
+          }
+       }
