=== modified file 'util/hostdisk.c'
--- util/hostdisk.c	2009-11-23 20:30:56 +0000
+++ util/hostdisk.c	2009-11-24 08:00:03 +0000
@@ -679,7 +679,7 @@ make_device_name (int drive, int dos_par
   char *bsd_part_str = NULL;
 
   if (dos_part >= 0)
-    asprintf (&dos_part_str, ",%d", dos_part + 1);
+    asprintf (&dos_part_str, ",%d", dos_part + (getenv ("GRUB_LEGACY_0_BASED_PARTITIONS") ? 0 : 1));
 
   if (bsd_part >= 0)
     asprintf (&bsd_part_str, ",%c", dos_part + 'a');

