Index: b/kern/emu/hostdisk.c
===================================================================
--- a/kern/emu/hostdisk.c
+++ b/kern/emu/hostdisk.c
@@ -1005,7 +1005,7 @@
   char *bsd_part_str = NULL;
 
   if (dos_part >= 0)
-    dos_part_str = xasprintf (",%d", dos_part + 1);
+    dos_part_str = xasprintf (",%d", dos_part + (getenv ("GRUB_LEGACY_0_BASED_PARTITIONS") ? 0 : 1));
 
   if (bsd_part >= 0)
     bsd_part_str = xasprintf (",%d", bsd_part + 1);
