Index: Makefile.in
===================================================================
--- Makefile.in	(revision 2402)
+++ Makefile.in	(working copy)
@@ -143,6 +143,8 @@
 	fi
 
 include $(srcdir)/conf/$(target_cpu)-$(platform).mk
+GRUB_EXTRAS=$(srcdir)/debian/grub-extras
+-include $(GRUB_EXTRAS)/conf/$(target_cpu)-$(platform).mk
 
 ### General targets.
 
Index: autogen.sh
===================================================================
--- autogen.sh	(revision 2402)
+++ autogen.sh	(working copy)
@@ -5,7 +5,7 @@
 autoconf
 autoheader
 echo timestamp > stamp-h.in
-for rmk in conf/*.rmk; do
+for rmk in conf/*.rmk debian/grub-extras/conf/*.rmk; do
   ruby genmk.rb < $rmk > `echo $rmk | sed 's/\.rmk$/.mk/'`
 done
 ./gendistlist.sh > DISTLIST
