diff -Nur -x '*.orig' -x '*~' grub2/util/grub.d/00_header.in grub2.new/util/grub.d/00_header.in
--- grub2/util/grub.d/00_header.in	2010-01-13 11:38:34.000000000 +0000
+++ grub2.new/util/grub.d/00_header.in	2010-01-13 11:45:07.000000000 +0000
@@ -57,6 +57,11 @@
     save_env saved_entry
   fi
 }
+
+function recordfail {
+  set recordfail=1
+  if [ -n \${have_grubenv} ]; then if [ -z \${boot_once} ]; then save_env recordfail; fi; fi
+}
 EOF
 
 case ${GRUB_TERMINAL_INPUT}:${GRUB_TERMINAL_OUTPUT} in
@@ -130,5 +135,9 @@
 fi
 
 cat << EOF
-set timeout=${GRUB_TIMEOUT}
+if [ \${recordfail} = 1 ]; then
+  set timeout=-1
+else
+  set timeout=${GRUB_TIMEOUT}
+fi
 EOF
diff -Nur -x '*.orig' -x '*~' grub2/util/grub.d/10_linux.in grub2.new/util/grub.d/10_linux.in
--- grub2/util/grub.d/10_linux.in	2010-01-13 11:38:34.000000000 +0000
+++ grub2.new/util/grub.d/10_linux.in	2010-01-13 11:38:36.000000000 +0000
@@ -70,6 +70,9 @@
     title="$(gettext "%s, with Linux %s")"
   fi
   printf "menuentry \"${title}\" {\n" "${os}" "${version}"
+  cat << EOF
+        recordfail
+EOF
   if [ "x$5" = "xquiet" ]; then
     cat << EOF
 	set quiet=1
