Ubuntu: Don't generate a device map by default.

diff -Nur -x '*.orig' -x '*~' no-device-map/util/grub-install.in no-device-map.new/util/grub-install.in
--- no-device-map/util/grub-install.in	2010-01-13 11:56:17.000000000 +0000
+++ no-device-map.new/util/grub-install.in	2010-01-13 11:57:33.000000000 +0000
@@ -38,7 +38,6 @@
 else
     grub_mkimage=${bindir}/`echo grub-mkelfimage | sed ${transform}`
 fi
-grub_mkdevicemap=${sbindir}/`echo grub-mkdevicemap | sed ${transform}`
 grub_probe=${sbindir}/`echo grub-probe | sed ${transform}`
 grub_editenv=${bindir}/`echo grub-editenv | sed ${transform}`
 rootdir=
@@ -73,7 +72,6 @@
                           instead of the root directory
   --grub-setup=FILE       use FILE as grub-setup
   --grub-mkimage=FILE     use FILE as grub-mkimage
-  --grub-mkdevicemap=FILE use FILE as grub-mkdevicemap
   --grub-probe=FILE       use FILE as grub-probe
   --no-floppy             do not probe any floppy drive
   --recheck               probe a device map even if it already exists
@@ -114,7 +112,7 @@
     --grub-mkimage=*)
 	grub_mkimage=`echo "$option" | sed 's/--grub-mkimage=//'` ;;
     --grub-mkdevicemap=*)
-	grub_mkdevicemap=`echo "$option" | sed 's/--grub-mkdevicemap=//'` ;;
+	: ;; # compatibility only
     --grub-probe=*)
 	grub_probe=`echo "$option" | sed 's/--grub-probe=//'` ;;
     --no-floppy)
@@ -199,14 +197,6 @@
     exit 1
 fi
 
-set $grub_mkdevicemap dummy
-if test -f "$1"; then
-    :
-else
-    echo "$1: Not found." 1>&2
-    exit 1
-fi
-
 # Create the GRUB directory if it is not present.
 test -d "$bootdir" || mkdir "$bootdir" || exit 1
 test -d "$grubdir" || mkdir "$grubdir" || exit 1
@@ -216,22 +206,14 @@
     rm -f $device_map
 fi
 
-# Create the device map file if it is not present.
+# Make sure that there is no duplicated entry in the device map.
 if test -f "$device_map"; then
-    :
-else
-    # Create a safe temporary file.
-    test -n "$mklog" && log_file=`$mklog`
-
-    $grub_mkdevicemap --device-map=$device_map $no_floppy || exit 1
-fi
-
-# Make sure that there is no duplicated entry.
-tmp=`sed -n '/^([fh]d[0-9]*)/s/\(^(.*)\).*/\1/p' $device_map \
-    | sort | uniq -d | sed -n 1p`
-if test -n "$tmp"; then
-    echo "The drive $tmp is defined multiple times in the device map $device_map" 1>&2
-    exit 1
+    tmp=`sed -n '/^([fh]d[0-9]*)/s/\(^(.*)\).*/\1/p' $device_map \
+	| sort | uniq -d | sed -n 1p`
+    if test -n "$tmp"; then
+	echo "The drive $tmp is defined multiple times in the device map $device_map" 1>&2
+	exit 1
+    fi
 fi
 
 # Copy the GRUB images to the GRUB directory.
