diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta4/util/grub.d/00_header.in grub2-1.97~beta4.new/util/grub.d/00_header.in
--- grub2-1.97~beta4/util/grub.d/00_header.in	2009-10-12 12:11:21.000000000 +0100
+++ grub2-1.97~beta4.new/util/grub.d/00_header.in	2009-10-12 12:11:22.000000000 +0100
@@ -112,5 +112,9 @@
 esac
 
 cat << EOF
-set timeout=${GRUB_TIMEOUT}
+if [ \${recordfail} = 1 ]; then
+  set timeout=-1
+else
+  set timeout=${GRUB_TIMEOUT}
+fi
 EOF
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta4/util/grub.d/10_linux.in grub2-1.97~beta4.new/util/grub.d/10_linux.in
--- grub2-1.97~beta4/util/grub.d/10_linux.in	2009-10-12 12:11:21.000000000 +0100
+++ grub2-1.97~beta4.new/util/grub.d/10_linux.in	2009-10-12 12:12:11.000000000 +0100
@@ -57,6 +57,8 @@
 {
   cat << EOF
 menuentry "$1" {
+        recordfail=1
+        if [ -n \${have_grubenv} ]; then save_env recordfail; fi
 EOF
   if [ "x$3" = "xquiet" ]; then
     cat << EOF
