--- grub2-1.97~beta4/util/grub.d/30_os-prober.in	2009-10-29 11:43:17.358148442 +0000
+++ grub2-1.97~beta4.new/util/grub.d/30_os-prober.in	2009-10-29 11:44:55.546922498 +0000
@@ -132,11 +132,16 @@
           LLABEL="${LONGNAME}"
         fi
 
+         if [ "${LROOT}" != "${LBOOT}" ]; then
+           LKERNEL="${LKERNEL#/boot}"
+           LINITRD="${LINITRD#/boot}"
+         fi
+
         cat << EOF
 menuentry "${LLABEL} (on ${DEVICE})" {
 EOF
 	save_default_entry | sed -e "s/^/\t/"
-	prepare_grub_to_access_device ${DEVICE} | sed -e "s/^/\t/"
+	prepare_grub_to_access_device ${LBOOT} | sed -e "s/^/\t/"
 	cat <<  EOF
 	linux ${LKERNEL} ${LPARAMS}
 EOF
