Ubuntu: fix compilation error
Upstream: backport of r2029 and r2030

diff -Nur -x '*.orig' -x '*~' grub2/util/mkisofs/write.c grub2.new/util/mkisofs/write.c
--- grub2/util/mkisofs/write.c	2010-01-05 10:07:49.000000000 +0000
+++ grub2.new/util/mkisofs/write.c	2010-01-05 16:12:28.000000000 +0000
@@ -1437,7 +1437,9 @@
       if (! fp)
 	error (1, errno, _("Unable to open %s"), boot_image_embed);
 
-      fread (buffer, 2048 * PADBLOCK_SIZE, 1, fp);
+      if (fread (buffer, 1, 2048 * PADBLOCK_SIZE, fp) == 0)
+	error (1, errno, _("cannot read %d bytes from %s"),
+	       2048 * PADBLOCK_SIZE, boot_image_embed);
       if (fgetc (fp) != EOF)
 	error (1, 0, _("%s is too big for embed area"), boot_image_embed);
     }
