Index: grub.zol/grub-core/fs/zfs/zfs.c
===================================================================
--- grub.zol.orig/grub-core/fs/zfs/zfs.c	2014-11-17 12:35:30.760500998 +0100
+++ grub.zol/grub-core/fs/zfs/zfs.c	2014-11-19 01:07:08.631348484 +0100
@@ -281,7 +281,10 @@
  */
 #define MAX_SUPPORTED_FEATURE_STRLEN 50
 static const char *spa_feature_names[] = {
-  "org.illumos:lz4_compress",NULL
+  "org.illumos:lz4_compress",
+  "com.delphix:hole_birth",
+  "com.delphix:embedded_data",
+  NULL
 };
 
 static int
