Description: Recognise DDF1 DM-RAID
Author: Modestas Vainius <modax@debian.org>
Origin: upstream, http://bazaar.launchpad.net/~vcs-imports/grub/grub2-bzr/revision/2946
Forwarded: yes
Last-Update: 2010-12-18

Index: b/kern/emu/getroot.c
===================================================================
--- a/kern/emu/getroot.c
+++ b/kern/emu/getroot.c
@@ -569,6 +569,8 @@
     return 1;
   else if (! strncmp (os_dev, "/dev/mapper/sil_", 16))
     return 1;
+  else if (! strncmp (os_dev, "/dev/mapper/ddf1_", 17))
+    return 1;
 
   return 0;
 }
