Description: Handle partitions on non-512B EFI disks
Origin: backport, http://bazaar.launchpad.net/~vcs-imports/grub/grub2-bzr/revision/4795
Author: Peter Jones <pjones@redhat.com>
Author: Vladimir Serbinenko <phcoder@gmail.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1065281
Forwarded: not-needed
Last-Update: 2013-09-18

Index: b/grub-core/disk/efi/efidisk.c
===================================================================
--- a/grub-core/disk/efi/efidisk.c
+++ b/grub-core/disk/efi/efidisk.c
@@ -668,9 +668,11 @@
 		&& (GRUB_EFI_DEVICE_PATH_SUBTYPE (c->last_device_path)
 		    == GRUB_EFI_HARD_DRIVE_DEVICE_PATH_SUBTYPE)
 		&& (grub_partition_get_start (disk->partition)
-		    == hd.partition_start)
+		    == (hd.partition_start << (disk->log_sector_size
+					       - GRUB_DISK_SECTOR_BITS)))
 		&& (grub_partition_get_len (disk->partition)
-		    == hd.partition_size))
+		    == (hd.partition_size << (disk->log_sector_size
+					      - GRUB_DISK_SECTOR_BITS))))
 	      {
 		handle = c->handle;
 		return 1;
@@ -763,11 +765,14 @@
       auto int find_partition (grub_disk_t disk, const grub_partition_t part);
 
       /* Find the identical partition.  */
-      int find_partition (grub_disk_t disk __attribute__ ((unused)),
-			  const grub_partition_t part)
+      int find_partition (grub_disk_t disk, const grub_partition_t part)
 	{
-	  if (grub_partition_get_start (part) == hd.partition_start
-	      && grub_partition_get_len (part) == hd.partition_size)
+	  if (grub_partition_get_start (part)
+	      == (hd.partition_start << (disk->log_sector_size
+					 - GRUB_DISK_SECTOR_BITS))
+	      && grub_partition_get_len (part)
+	      == (hd.partition_size << (disk->log_sector_size
+					- GRUB_DISK_SECTOR_BITS)))
 	    {
 	      partition_name = grub_partition_get_name (part);
 	      return 1;
@@ -799,7 +804,9 @@
       /* Find a partition which matches the hard drive device path.  */
       grub_memcpy (&hd, ldp, sizeof (hd));
       if (hd.partition_start == 0
-	  && hd.partition_size == grub_disk_get_size (parent))
+	  && (hd.partition_size << (parent->log_sector_size
+				    - GRUB_DISK_SECTOR_BITS))
+	      == grub_disk_get_size (parent))
 	{
 	  dev_name = grub_strdup (parent->name);
 	}
