Index: grub/grub-core/fs/zfs/zfs.c
===================================================================
--- grub.orig/grub-core/fs/zfs/zfs.c	2015-01-14 11:44:46.768454805 +0100
+++ grub/grub-core/fs/zfs/zfs.c	2015-01-14 11:48:15.297833231 +0100
@@ -281,7 +281,16 @@
  */
 #define MAX_SUPPORTED_FEATURE_STRLEN 50
 static const char *spa_feature_names[] = {
-  "org.illumos:lz4_compress",NULL
+  "org.illumos:lz4_compress",
+  "com.delphix:hole_birth",
+  "com.delphix:embedded_data",
+  "spacemap_histogram",
+  "enabled_txg",
+  "extensible_dataset",
+  "bookmarks",
+  "async_destroy",
+  "empty_bpobj",
+  NULL
 };
 
 static int
