Index: grub/tests/util/grub-shell.in
===================================================================
--- grub.orig/tests/util/grub-shell.in	2015-03-18 13:19:59.405828522 +0100
+++ grub/tests/util/grub-shell.in	2015-03-18 13:20:34.825383362 +0100
@@ -375,8 +375,8 @@
     bootdev="-boot d"
 fi
 if [ x$boot = xfd ]; then
-    device="fda "
-    bootdev="-boot a"
+    device="cdrom "
+    bootdev="-boot d"
 fi
 
 if [ x$boot = xqemu ]; then
