Index: grub/tests/util/grub-fs-tester.in
===================================================================
--- grub.orig/tests/util/grub-fs-tester.in	2015-03-17 12:58:06.960274531 +0100
+++ grub/tests/util/grub-fs-tester.in	2015-03-17 12:58:51.583713537 +0100
@@ -426,6 +426,8 @@
 		x"vfat"* | xmsdos* | x"hfs" | x"exfat" | x"ntfs"* \
 		    | xiso9660 | xjoliet| xiso9660_1999 | xjoliet_1999)
 		    NOSYMLINK=y;;
+		x"zfs"*)
+		    NOSYMLINK=y;;
 	    esac
 	    NOHARDLINK=n
 	    case x"$fs" in
@@ -480,6 +482,8 @@
 	# FS LIMITATION: romfs has no timestamps.
 		x"romfs")
 		    NOFILETIME=y; NOFSTIME=y;;
+		x"zfs"*)
+		    NOFILETIME=y; NOFSTIME=y;;
 	    esac
 
 	    NOFSLABEL=n
@@ -843,7 +847,7 @@
 	    GRUBDIR="($GRUBDEVICE)"
 	    case x"$fs" in
 		x"zfs"*)
-		    OSDIR="grub fs/"
+		    OSDIR="$FSLABEL/grub fs/"
 		    GRUBDIR="($GRUBDEVICE)/grub fs@";;
 		x"tarfs" | x"cpio_"* | x"iso9660" | xjoliet | xrockridge | xrockridge_joliet | x"iso9660_1999" | xjoliet_1999 | xrockridge_1999 | xrockridge_joliet_1999 | x"ziso9660" | x"romfs" | x"squash4_"* | xafs)
 		    ;;
