Index: grub/tests/util/grub-shell.in
===================================================================
--- grub.orig/tests/util/grub-shell.in	2015-03-23 16:54:53.269536444 +0100
+++ grub/tests/util/grub-shell.in	2015-03-23 16:55:00.481445781 +0100
@@ -376,8 +376,8 @@
     bootdev="-boot d"
 fi
 if [ x$boot = xfd ]; then
-    device="fda "
-    bootdev="-boot a"
+    device="cdrom "
+    bootdev="-boot d"
 fi
 
 if [ x$boot = xqemu ]; then
