Description: Add support for ext2 root on GNU/kFreeBSD
Author: Aurelien Jarno <aurel32@debian.org>
Bug-Debian: http://bugs.debian.org/593467
Forwarded: http://lists.gnu.org/archive/html/grub-devel/2010-08/msg00038.html
Last-Update: 2010-08-19

Index: b/util/grub.d/10_kfreebsd.in
===================================================================
--- a/util/grub.d/10_kfreebsd.in
+++ b/util/grub.d/10_kfreebsd.in
@@ -75,6 +75,14 @@
   fi
 
   case "${kfreebsd_fs}" in
+    ext2fs)
+      test -e "${module_dir}/ext2fs.ko"
+
+      printf '%s\n' "${prepare_module_dir_cache}"
+      cat << EOF
+	kfreebsd_module_elf	${module_dir_rel}/ext2fs.ko
+EOF
+    ;;
     zfs)
       test -e "${module_dir}/opensolaris.ko"
       test -e "${module_dir}/zfs.ko"
@@ -121,6 +129,7 @@
 
   case ${GRUB_FS} in
     ufs1 | ufs2)        kfreebsd_fs=ufs ;;
+    ext2)               kfreebsd_fs=ext2fs ;;
     *)                  kfreebsd_fs=${GRUB_FS} ;;
   esac
 
