Description: "single" -> "recovery" when friendly-recovery is installed
Author: Colin Watson <cjwatson@ubuntu.com>
Author: Stéphane Graber <stgraber@ubuntu.com>
Forwarded: no
Last-Update: 2013-01-21

Index: b/util/grub.d/10_linux.in
===================================================================
--- a/util/grub.d/10_linux.in
+++ b/util/grub.d/10_linux.in
@@ -243,8 +243,13 @@
   linux_entry "${OS}" "${version}" advanced \
               "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
   if [ "x${GRUB_DISABLE_RECOVERY}" != "xtrue" ]; then
-    linux_entry "${OS}" "${version}" recovery \
-                "single ${GRUB_CMDLINE_LINUX}"
+    if [ -x /lib/recovery-mode/recovery-menu ]; then
+      linux_entry "${OS}" "${version}" recovery \
+                  "recovery ${GRUB_CMDLINE_LINUX}"
+    else
+      linux_entry "${OS}" "${version}" recovery \
+                  "single ${GRUB_CMDLINE_LINUX}"
+    fi
   fi
 
   list=`echo $list | tr ' ' '\n' | grep -vx $linux | tr '\n' ' '`
Index: b/util/grub.d/30_os-prober.in
===================================================================
--- a/util/grub.d/30_os-prober.in
+++ b/util/grub.d/30_os-prober.in
@@ -170,7 +170,7 @@
 	fi
 
 	onstr="$(gettext_printf "(on %s)" "${DEVICE}")"
-	recovery_params="$(echo "${LPARAMS}" | grep single)" || true
+	recovery_params="$(echo "${LPARAMS}" | grep 'single\|recovery')" || true
 	counter=1
 	while echo "$used_osprober_linux_ids" | grep 'osprober-gnulinux-$LKERNEL-${recovery_params}-$counter-$boot_device_id' > /dev/null; do
 	    counter=$((counter+1));
