Description: Initialise next pointer when creating multiboot module
Author: Vladimir Serbinenko <phcoder@gmail.com>
Origin: upstream, http://bazaar.launchpad.net/~vcs-imports/grub/grub2-bzr/revision/2986
Bug-Debian: http://bugs.debian.org/605567
Forwarded: yes
Last-Update: 2010-12-21

Index: b/loader/i386/multiboot_mbi.c
===================================================================
--- a/loader/i386/multiboot_mbi.c
+++ b/loader/i386/multiboot_mbi.c
@@ -465,6 +465,7 @@
     return grub_errno;
   newmod->start = start;
   newmod->size = size;
+  newmod->next = 0;
 
   for (i = 0; i < argc; i++)
     len += grub_strlen (argv[i]) + 1;
