Description: Probe FusionIO devices
Author: Colin Watson <cjwatson@ubuntu.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1237519
Forwarded: no
Last-Update: 2013-11-13

Index: b/util/deviceiter.c
===================================================================
--- a/util/deviceiter.c
+++ b/util/deviceiter.c
@@ -365,6 +365,12 @@
 {
   sprintf (name, "/dev/xvd%c", unit + 'a');
 }
+
+static void
+get_fio_disk_name (char *name, int unit)
+{
+  sprintf (name, "/dev/fio%c", unit + 'a');
+}
 #endif
 
 static struct seen_device
@@ -852,6 +858,19 @@
       if (check_device_readable_unique (name))
 	{
 	  if (hook (name, 0))
+	    goto out;
+	}
+    }
+
+  /* FusionIO.  */
+  for (i = 0; i < 26; i++)
+    {
+      char name[16];
+
+      get_fio_disk_name (name, i);
+      if (check_device_readable_unique (name))
+	{
+	  if (hook (name, 0))
 	    goto out;
 	}
     }
Index: b/util/getroot.c
===================================================================
--- a/util/getroot.c
+++ b/util/getroot.c
@@ -1825,6 +1825,19 @@
 	  return path;
 	}
 
+      /* If this is a FusionIO disk.  */
+      if ((strncmp ("fio", p, 3) == 0) && p[3] >= 'a' && p[3] <= 'z')
+	{
+	  char *pp = p + 3;
+	  while (*pp >= 'a' && *pp <= 'z')
+	    pp++;
+	  if (*pp)
+	    *is_part = 1;
+	  /* /dev/fio[a-z]+[0-9]* */
+	  *pp = '\0';
+	  return path;
+	}
+
 #ifdef HAVE_DEVICE_MAPPER
       if ((strncmp ("/dev/mapper/", path, sizeof ("/dev/mapper/") - 1) == 0)
 	  || (strncmp ("/dev/dm-", path, sizeof ("/dev/dm-") - 1) == 0))
