Description: Fix grub-mkrelpath for non-root ZFS
Author: Robert Millan <rmh@debian.org>
Forwarded: not needed, patch from upstream Bazaar
Last-Update: 2010-10-18

--- a/kern/emu/misc.c	2010-08-03 22:15:29 +0000
+++ b/kern/emu/misc.c	2010-10-18 10:51:32 +0000
@@ -456,21 +456,7 @@ grub_make_system_path_relative_to_its_ro
 
       /* buf is another filesystem; we found it.  */
       if (st.st_dev != num)
-	{
-	  /* offset == 0 means path given is the mount point.
-	     This works around special-casing of "/" in Un*x.  This function never
-	     prints trailing slashes (so that its output can be appended a slash
-	     unconditionally).  Each slash in is considered a preceding slash, and
-	     therefore the root directory is an empty string.  */
-	  if (offset == 0)
-	    {
-	      free (buf);
-	      free (buf2);
-	      return xstrdup ("");
-	    }
-	  else
-	    break;
-	}
+	break;
 
       offset = p - buf;
       /* offset == 1 means root directory.  */

