Description: Accept ZFS up to v23 (no changes required)
Author: Robert Millan <rmh@debian.org>
Forwarded: not needed, patch from upstream Bazaar
Last-Update: 2010-09-22

--- a/debian/grub-extras/zfs/include/grub/zfs/zfs.h	2009-10-26 20:15:26 +0000
+++ b/debian/grub-extras/zfs/include/grub/zfs/zfs.h	2010-09-22 15:43:24 +0000
@@ -31,7 +31,7 @@
 /*
  * On-disk version number.
  */
-#define	SPA_VERSION			14ULL
+#define	SPA_VERSION			23ULL
 
 /*
  * The following are configuration names used in the nvlist describing a pool's

