Description: Fix pathname for non-root ZFS filesystems
Author: Robert Millan <rmh@debian.org>
Forwarded: not needed, patch from upstream Bazaar
Last-Update: 2010-08-16

Index: b/kern/emu/misc.c
===================================================================
--- a/kern/emu/misc.c
+++ b/kern/emu/misc.c
@@ -519,7 +519,7 @@
 
   if (poolfs)
     {
-      ret = xasprintf ("/%s@%s", poolfs, buf3);
+      ret = xasprintf ("/%s/@%s", poolfs, buf3);
       free (buf3);
     }
   else
