Index: b/grub-core/fs/zfs/zfs.c
===================================================================
--- a/grub-core/fs/zfs/zfs.c
+++ b/grub-core/fs/zfs/zfs.c
@@ -377,13 +377,11 @@
   zio_cksum_t zc;
 
   if (grub_zfs_to_cpu64 (uber->ub_magic, LITTLE_ENDIAN) == UBERBLOCK_MAGIC
-      && grub_zfs_to_cpu64 (uber->ub_version, LITTLE_ENDIAN) > 0 
-      && grub_zfs_to_cpu64 (uber->ub_version, LITTLE_ENDIAN) <= SPA_VERSION)
+      && SPA_VERSION_IS_SUPPORTED(grub_zfs_to_cpu64 (uber->ub_version,LITTLE_ENDIAN)))
     endian = LITTLE_ENDIAN;
 
   if (grub_zfs_to_cpu64 (uber->ub_magic, BIG_ENDIAN) == UBERBLOCK_MAGIC
-      && grub_zfs_to_cpu64 (uber->ub_version, BIG_ENDIAN) > 0 
-      && grub_zfs_to_cpu64 (uber->ub_version, BIG_ENDIAN) <= SPA_VERSION)
+      && SPA_VERSION_IS_SUPPORTED(grub_zfs_to_cpu64 (uber->ub_version,BIG_ENDIAN)))
     endian = BIG_ENDIAN;
 
   if (endian == UNKNOWN_ENDIAN)
@@ -673,13 +671,13 @@
     }
   grub_dprintf ("zfs", "check 8 passed\n");
 
-  if (version > SPA_VERSION)
+  if (!SPA_VERSION_IS_SUPPORTED(version))
     {
       grub_free (nvlist);
       return grub_error (GRUB_ERR_NOT_IMPLEMENTED_YET,
 			 "too new version %llu > %llu",
 			 (unsigned long long) version,
-			 (unsigned long long) SPA_VERSION);
+			 (unsigned long long) SPA_VERSION_BEFORE_FEATURES);
     }
   grub_dprintf ("zfs", "check 9 passed\n");
 
Index: b/include/grub/zfs/zfs.h
===================================================================
--- a/include/grub/zfs/zfs.h
+++ b/include/grub/zfs/zfs.h
@@ -28,7 +28,13 @@
 /*
  * On-disk version number.
  */
-#define	SPA_VERSION			33ULL
+#define SPA_VERSION_INITIAL		1ULL
+#define SPA_VERSION_BEFORE_FEATURES	33ULL
+#define SPA_VERSION			5000ULL
+#define SPA_VERSION_FEATURES		5000ULL
+#define SPA_VERSION_IS_SUPPORTED(v) \
+  (((v) >= SPA_VERSION_INITIAL && (v) <= SPA_VERSION_BEFORE_FEATURES) || \
+  ((v) >= SPA_VERSION_FEATURES && (v) <= SPA_VERSION))
 
 /*
  * The following are configuration names used in the nvlist describing a pool's
