Description: Fix grub-mkrelpath for non-root ZFS
Author: Vladimir 'phcoder' Serbinenko
Forwarded: not needed, patch from upstream Bazaar (r2905)
Last-Update: 2010-10-24

--- a/kern/emu/misc.c	2010-10-23 18:34:50 +0000
+++ b/kern/emu/misc.c	2010-10-23 18:39:08 +0000
@@ -416,6 +416,10 @@ grub_make_system_path_relative_to_its_ro
 	    {
 	      free (buf);
 	      free (buf2);
+#if defined(HAVE_LIBZFS) && defined(HAVE_LIBNVPAIR)
+	      if (poolfs)
+		return xasprintf ("/%s/@", poolfs);
+#endif
 	      return xstrdup ("");
 	    }
 	  else

