Description: Fix breakage when running grub-probe inside chroot
Author: Robert Millan <rmh@debian.org>
Forwarded: not needed, patch from upstream Bazaar
Last-Update: 2010-08-16

--- a/kern/emu/misc.c	2010-08-03 22:15:29 +0000
+++ b/kern/emu/misc.c	2010-08-16 13:07:46 +0000
@@ -344,7 +345,17 @@ grub_find_zpool_from_mount_point (const
 
   *poolname = *poolfs = NULL;
 
-#if defined(HAVE_GETFSSTAT) /* FreeBSD and GNU/kFreeBSD */
+#if 1
+  {
+    struct statfs mnt;
+
+    if (statfs (mnt_point, &mnt) != 0)
+      error (1, errno, "statfs");
+
+    if (!strcmp (mnt.f_fstypename, "zfs"))
+      *poolname = xstrdup (mnt.f_mntfromname);
+  }
+#elif defined(HAVE_GETFSSTAT) /* FreeBSD and GNU/kFreeBSD */
   {
     int mnt_count = getfsstat (NULL, 0, MNT_WAIT);
     if (mnt_count == -1)
