Description: Fix grub-probe fs_label argument
Author: Robert Millan <rmh@debian.org>
Forwarded: not needed, patch from upstream Bazaar
Last-Update: 2010-08-16

--- a/util/grub.d/10_kfreebsd.in	2010-08-01 21:21:09 +0000
+++ b/util/grub.d/10_kfreebsd.in	2010-08-16 13:07:46 +0000
@@ -125,7 +125,7 @@ while [ "x$list" != "x" ] ; do
   esac
 
   case ${GRUB_FS} in
-    zfs)		kfreebsd_device=$(grub-probe -t label --device ${GRUB_DEVICE}) ;;
+    zfs)		kfreebsd_device=$(grub-probe -t fs_label --device ${GRUB_DEVICE}) ;;
     *)			kfreebsd_device=${GRUB_DEVICE} ;;
   esac
 

